All the derives, add fallible point conversions

jamwaffles opened the pull request:
Hi! Thank you for helping out with Embedded Graphics development! Please:

- [x] Check that you've added passing tests and documentation
- [x] Add a simulator example(s) where applicable
- [x] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc)
- [x] Run `rustfmt` on the project
- [x] Run `./build.sh` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

TODO

- [x] Add a shit load of derives. This was a find and replace with some manual fixes where some derives couldn't be used, so some might not make sense. Closes #238.
- [x] Add fallible conversions for `(u32, u32)` and `Point`. Closes #242.
- [x] Revert dependency fudging (due to alpha version numbers in `Cargo.toml`s I think)
- [x] ~Decide whether to add the following~ - not doing it. It can be added later if it's a common usage pattern.

  ```rust
  impl TryFrom<Size> for (i32, i32) { ... }
  impl TryFrom<(i32, i32)> for Size { ... }
  ```
- [x] https://github.com/jamwaffles/embedded-graphics/pull/243#discussion_r367593226

jamwaffles commented on embedded-graphics/Cargo.toml:
```diff
@@ -26,8 +26,8 @@ circle-ci = { repository = "jamwaffles/embedded-graphics", branch = "master" }
-tinybmp = { version = "0.1.0", optional = true }
-tinytga = { version = "0.2.0", optional = true }
+tinybmp = { version = "0.1.1-alpha.0", optional = true }
+tinytga = { version = "0.2.1-alpha.0", optional = true }
```
These need to be reverted before merge.

jamwaffles commented:
I took a stab at https://github.com/jamwaffles/embedded-graphics/pull/243#discussion_r367593226 but I'm not too keen on the examples. They do show what's going on, but they feel quite noisy with the added `DisplayError` enum. Any thoughts on a better approach (if necessary)?

rfuest commented:
I thought we agreed that drawing out of bounds shouldn't be an error, but a noop instead. Otherwise code like this:
```rust
fn draw_something(display: &mut Display) -> Result<(), Err> {
   SomeDrawable::new(...).draw(display)?;
   OtherDrawable::new(...).draw(display)
}
```
wouldn't even try to draw the second drawable if `SomeDrawable` overlaps the display area.

jamwaffles commented:
Right, yeah we did. I misunderstood what you meant by https://github.com/jamwaffles/embedded-graphics/pull/243#discussion_r367538339 then. In that case,

1. What should the error type of the examples be? `Infallible`? Keep the custom enum?
2. `draw_pixel()` could be changed to use `unwrap()`:

    ```rust
    let Pixel(coord, color) = pixel;
    
    if coord.x >= 0 && coord.x < 64 && coord.y >= 0 && coord.y < 64 {
        let x = usize::try_from(coord.x).unwrap();
        let y = usize::try_from(coord.y).unwrap();
    
        let index: usize = x + y * 64;
    
        self.framebuffer[index] = color.luma();
    }
  
    Ok(())
    ```
  
    or just use `as` because it'll never fail:
  
    ```rust
    let Pixel(coord, color) = pixel;
    
    if coord.x >= 0 && coord.x < 64 && coord.y >= 0 && coord.y < 64 {
    	  // Coercion from `i32` to `usize` here cannot fail due to bounds check above
        let index: usize = x as usize + y as usize * 64;
    
        self.framebuffer[index] = color.luma();
    }
  
    Ok(())
    ```
  
    I could also just revert the examples back to what's in `master`.

Thoughts?

rfuest commented:
>     1. What should the error type of the examples be? `Infallible`? Keep the custom enum?

Having at least one example that shows how and when to use an error type other that `Infallible` would be useful. My suggestion would be to change the first example back to using `Infallible` and remove the framebuffer from the hardware accelerated example. This way every draw operation in the second example would need to communicate with the display and could fail. The `Error` type could be something like `SpiError` in this case.

> I misunderstood what you meant by [#243 (comment)](https://github.com/jamwaffles/embedded-graphics/pull/243#discussion_r367538339) then.

What I meant is that if #242 is implemented `draw_pixel` could be implemented as:
```rust
let Pixel(coord, color) = pixel;

if let Some((x, y)) = coord.try_into() {
    if x < 64 && y < 64 {
        let index: u32 = x + y * 64;
        self.framebuffer[index as usize] = color.luma();
    }
}

Ok(())
```
or:
```rust
let Pixel(coord, color) = pixel;

if let Some((x @ 0..=63, y @ 0..=63)) = coord.try_into() {
    let index: u32 = x + y * 64;
    self.framebuffer[index as usize] = color.luma();
}

Ok(())
```

jamwaffles commented:
Ahh I see now. TIL the `x @ 0..=63` syntax exists - that's awesome! I've updated the `DrawTarget` examples. Might need some doc tweaks as usual 🙂

rfuest commented:
Is this ready for review? I'm asking because the review request is still open from before the last changes.

jamwaffles commented:
Yep! Sorry if that wasn’t clear in the comment.

rfuest commented on embedded-graphics/src/geometry/point.rs:
```diff
@@ -279,6 +299,41 @@ where
+
+        let tuple: (u32, u32) = p.try_into().unwrap();
+
+        // Literal value taken from [here](https://doc.rust-lang.org/std/primitive.i32.html#method.max_value)
```
Because this isn't a doc comment there is no benefit of using the markdown syntax here IMO.

rfuest commented on embedded-graphics/src/geometry/point.rs:
```diff
@@ -279,6 +299,41 @@ where
+    }
+
+    #[test]
+    #[should_panic]
```
This test and the next one shouldn't use `unwrap` together with the `should_panic` attribute, because this way the tests would also pass if, for example, `Point::new` panics. Explicitly testing the return value of `try_into` would remove this possibility for false positives.

jamwaffles replied:
I first tried an `assert_eq!(tuple, Err(core::num::TryFromIntError))` but `TryFromIntError` can't be constructed. I went with `assert!(tuple.is_err())` instead. Is there a better check?

rfuest replied:
> Is there a better check?

IMO `assert!(tuple.is_err())` is a good check in this case. There is only one possible error type which doesn't have any parameters, so that both asserts in your comment are equivalent.

jamwaffles replied:
Sweet 👌

rfuest commented on embedded-graphics/src/geometry/point.rs:
```diff
@@ -279,6 +299,41 @@ where
+
+        let tuple: (u32, u32) = p.try_into().unwrap();
+
+        assert_eq!(tuple, (10u32, 20u32));
```
Nit: The type annotation isn't necessary.

jamwaffles commented:
Comments addressed 👍

rfuest commented:
> Comments addressed

These changes look good. Did you also notice the comment about `TryFrom` for arrays and slices?

jamwaffles commented:
> Did you also notice the comment about TryFrom for arrays and slices?

Oop, missed that. Do you mean `impl TryFrom<[u32; 2]> for Point` and `impl TryFrom<&[u32; 2]> for Point`?

rfuest commented:
> Do you mean `impl TryFrom<[u32; 2]> for Point` and `impl TryFrom<&[u32; 2]> for Point`?

Yes, that's what I meant. We did decide to keep the `From` array conversions while renaming `Coord` to `Point`. Adding the same kind of conversions for `u32` arrays would improve consistency.

I also just noticed that the opposite conversion `From<Point> for [i32; 2]` doesn't exist. Should we add these too (also for `TryFrom`)?

jamwaffles commented:
> Yes, that's what I meant.

Cool, added. I doubled up the tests at the bottom of `point.rs` instead of adding new array-specific ones as the file is already hella long.

> I also just noticed that the opposite conversion From<Point> for [i32; 2] doesn't exist. Should we add these too (also for TryFrom)?

Makes sense to me, so I've added it.

I gotta be honest all these conversion impls are frying my brain, so apologies in advance for any stupid mistakes/omissions. I'm starting to think about impl macros to make life easier, but not in this PR :)

rfuest commented:
Great, I'll review the changes together with the derives, but this will probably have to wait until next week.

> I gotta be honest all these conversion impls are frying my brain, so apologies in advance for any stupid mistakes/omissions. I'm starting to think about impl macros to make life easier, but not in this PR :)

No problem, I can understand that and I'm not looking forward to reviewing the code :wink:. Should you have some motivation left you could also add `impl From<Size> for [u32; 2]`.

jamwaffles commented:
> this will probably have to wait until next week.

No problem, thanks for the time spent on all these reviews. I should note that the derive additions were very much a copypasta all the derives everywhere, then remove anything that errored in the compiler. There might be some derives lying around that don't make sense.

> Should you have some motivation left you could also add impl From<Size> for [u32; 2].

Good idea, added.

jamwaffles commented:
I realise there's a lot of code to review @rfuest but if you have time could you look through this too?

rfuest commented:
> I realise there's a lot of code to review @rfuest but if you have time could you look through this too?

Yes, I hope that I'll find some spare time to do the code reviews on the weekend.

jamwaffles commented on embedded-graphics/src/style/primitive_style.rs:
```diff
@@ -14,7 +14,7 @@ use core::convert::TryFrom;
 /// [`non_exhaustive`]: https://blog.rust-lang.org/2019/12/19/Rust-1.40.0.html#[non_exhaustive]-structs,-enums,-and-variants
 /// [`primitive_style!`]: ../macro.primitive_style.html
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
```
Replacing the manual impl with a derive leads to some odd results. I had to pepper some extra `+ Default` trait bounds on `C` and it got messy fast so I gave up. I've swapped the `new()` and `default()` method bodies as I think that makes a bit more sense - `PrimitiveStyle::new()` now calls `PrimitiveStyle::default()` instead of the other way round. Not a huge change either way tbh.

rfuest replied:
Interesting, I thought deriving would work because `Default` is implemented for every `Option<T>` and not only `Option<T>`s where `T: Default`.
I only suggested this as a minor cleanup and I'm OK with leaving the code the way it is now.

jamwaffles commented on embedded-graphics/src/draw_target.rs:
```diff
@@ -100,8 +107,12 @@ use crate::{
 /// alternative implementations for hardware accelerated drawing operations. This example implements
-/// `DrawTarget` for a display that supports drawing hardware accelerated styled [`Rectangle`]s. The
-/// `draw_rectangle()` method overrides the default implementation.
+/// `DrawTarget` for a display without a framebuffer that supports drawing hardware accelerated
```
Take a look at this whole example intro section. I've split it into 3 paragraphs which are hopefully a bit clearer.

jamwaffles commented on embedded-graphics/src/draw_target.rs:
```diff
@@ -100,8 +107,12 @@ use crate::{
-/// `DrawTarget` for a display that supports drawing hardware accelerated styled [`Rectangle`]s. The
-/// `draw_rectangle()` method overrides the default implementation.
+/// `DrawTarget` for a display without a framebuffer that supports drawing hardware accelerated
+/// styled [`Rectangle`]s. The `draw_rectangle()` method overrides the default implementation.
```
See above comment

jamwaffles commented on embedded-graphics/src/draw_target.rs:
```diff
@@ -100,8 +107,12 @@ use crate::{
+///
+/// As this example doesn't use a framebuffer, a `flush()` method is not required. All draw
+/// operations are performed in "immediate mode" directly on the display. As drawing operations are
+/// now fallible due to hardware issues, a custom error type `CommError` is introduced.
```
See above comment

jamwaffles commented on embedded-graphics/src/image/image_raw.rs:
```diff
@@ -93,7 +93,7 @@ pub type ImageBE<'a, C> = Image<'a, C, BigEndian>;
 /// [`PixelColor`]: ../pixelcolor/trait.PixelColor.html
 /// [`ByteOrder`]: ../pixelcolor/raw/trait.ByteOrder.html
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
```
Nah it doesn't, I've removed it.

jamwaffles commented on embedded-graphics/src/pixelcolor/binary_color.rs:
````diff
@@ -39,7 +39,7 @@ use crate::pixelcolor::{
 /// assert_eq!(color, BinaryColor::On);
 /// ```
-#[derive(Debug, Clone, Copy, PartialEq)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
````
Makes sense, added with a test.

jamwaffles commented:
Cheers for the review. I've removed nearly all the `Default` derives you picked out - the blanket copypasta was a bit overreaching and most of them certainly don't make sense. Ready for another pass when you're free.

rfuest commented on embedded-graphics/src/draw_target.rs:
```diff
@@ -100,8 +107,18 @@ use crate::{
 ///
 /// In addition to defining [`draw_pixel`], an implementation of [`DrawTarget`] can also provide
 /// alternative implementations for hardware accelerated drawing operations. This example implements
```
This might be misleading, because "alternative implementations" could be interpreted as alternative implementations of `draw_pixel`.

jamwaffles replied:
Yeah maybe it's a bit vague, but IMO reading on to the next paragraphs and looking at the code example clarifies things. Can we leave this as is for now? Docs are a simple addition to future releases. Non breaking changes from e-g? Who'd've thought???!!1/!?!?1/1/1 :grin:

rfuest replied:
> Can we leave this as is for now?

OK.

> Non breaking changes from e-g? Who'd've thought???!!1/!?!?1/1/1 grin

That's a novel idea, but I promise I'll find more ways to break the code in future releases :wink:

jamwaffles replied:
> That's a novel idea

I'm gonna patent it

> I promise I'll find more ways to break the code in future releases 

Wahey but that's what semver pre-1.0.0 is for :+1:

rfuest commented on embedded-graphics/src/draw_target.rs:
```diff
@@ -100,8 +107,18 @@ use crate::{
+/// The default implementation of `draw_rectangle()` (as well as `draw_line()`, `draw_circle()` and
+/// `draw_triangle()`) defer to `draw_iter()` internally. In this example, the default
+/// implementation of `draw_rectangle()` is overridden to allow usage of accelerated draw commands
+/// specific to the targeted hardware.
```
IMO the function references should be links and the `()` removed.
Explicitly listing all `draw_...` functions increases the risk of this comment to become out of sync with the code, when additional methods are added.

jamwaffles replied:
Good shout. I've tweaked it a bit. I think the intent is pretty clear now so let's leave this doc section as is.

rfuest commented on embedded-graphics/src/draw_target.rs:
```diff
@@ -100,8 +107,18 @@ use crate::{
+/// implementation of `draw_rectangle()` is overridden to allow usage of accelerated draw commands
+/// specific to the targeted hardware.
+///
+/// As this example doesn't use a framebuffer, a `flush()` method is not required. All draw
```
The way the framebuffer is flushed and the name of the method isn't relevant here. The previous example uses the more generic `"flush" operation` which could also be used here.

jamwaffles replied:
Tweaked as well, thanks for the suggestion

rfuest commented on embedded-graphics/src/primitives/circle.rs:
````diff
@@ -51,7 +51,7 @@ use crate::{
 /// # Ok::<(), core::convert::Infallible>(())
 /// ```
-#[derive(Debug, Copy, Clone, PartialEq)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
````
```suggestion
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
```

rfuest commented on embedded-graphics/src/primitives/circle.rs:
```diff
@@ -122,7 +122,7 @@ impl Transform for Circle {
 
 /// Pixel iterator for each pixel in the circle border
-#[derive(Debug, Copy, Clone)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
```
```suggestion
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
```

rfuest commented on embedded-graphics/src/primitives/line.rs:
````diff
@@ -37,7 +37,7 @@ use crate::{
 /// # Ok::<(), core::convert::Infallible>(())
 /// ```
-#[derive(Debug, Copy, Clone, PartialEq)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
````
```suggestion
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
```

rfuest commented on embedded-graphics/src/primitives/line.rs:
```diff
@@ -149,7 +149,7 @@ where
 
 /// Pixel iterator for each pixel in the line
-#[derive(Debug, Clone, Copy)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
```
```suggestion
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
```

rfuest commented on embedded-graphics/src/primitives/rectangle.rs:
````diff
@@ -43,7 +43,7 @@ use crate::{
 /// # Ok::<(), core::convert::Infallible>(())
 /// ```
-#[derive(Debug, Clone, Copy, PartialEq)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
````
```suggestion
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
```

rfuest commented on embedded-graphics/src/primitives/rectangle.rs:
```diff
@@ -136,7 +136,7 @@ where
 
 /// Pixel iterator for each pixel in the rect border
-#[derive(Debug, Clone, Copy)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
```
```suggestion
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
```

rfuest commented on embedded-graphics/src/primitives/triangle.rs:
````diff
@@ -60,7 +60,7 @@ use core::borrow::Borrow;
 /// # assert_eq!(tri_ref, Triangle::new(p1, p2, p3));
 /// ```
-#[derive(Debug, Clone, Copy, PartialEq)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
````
```suggestion
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
```

rfuest commented on embedded-graphics/src/primitives/triangle.rs:
```diff
@@ -221,7 +221,7 @@ enum IterState {
 
 /// Pixel iterator for each pixel in the triangle border
-#[derive(Debug, Clone, Copy)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
```
```suggestion
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
```

jamwaffles commented:
Oh! From the _iterator_, not the primitive itself... you can blame Github's lack of context in review comments on the main thread for that lol. Fixed along with other suggestions.

rfuest commented:
LGTM :+1:

jamwaffles commented:
Beaut! Thank you!
